EpiAware/src/EpiInfModels/ODEProcess.jl

```diff
@@ -0,0 +1,253 @@
+    initial_prop_infected = Beta(1, 99)
+)
+```
+
+# Usage example with `ODEProcess` and predefined SIR model
+
+In this example we define an `ODEProcess` object using the predefined `SIRParams` model from
+above. We then generate latent infections using the `generate_latent_infs` function, and refit
+the model using a `Turing` model.
+
+We assume that the latent infections are observed with a Poisson likelihood around their
+ODE model prediction. The population size is `N = 1000`, which we put into the `sol2infs` function,
+which maps the ODE solution to the number of infections. Recall that the `EpiAware` default SIR
+implementation assumes the model is in density/proportion form. Also, note that since the `sol2infs`
+function is a link function that maps the ODE solution to the expected number of infections we also
+apply the `LogExpFunctions.logaddexp` function to ensure that the expected number of infections is non-negative.
+The `logaddexp` function generalises the `softplus` function, `logaddexp(1, x) == softplus(x)`.
+
+First, we define the `ODEProcess` object which combines the SIR model with the `sol2infs` link
+function and the solver options.
+
+```julia
+using Turing, LogExpFunctions
+N = 1000.0
+
+sir_process = ODEProcess(
+    params = sirparams,
+    sol2infs = sol -> N .* logaddexp.(0.001, sol[2, :]),
+    solver_options = Dict(:verbose => false, :saveat => 1.0)
+)
+```
+
+Next, we create a `Turing` model for the infection generating process using the `generate_latent_infs` function.
+This function remakes the ODE problem with the provided initial conditions and parameters, solves it using the
+specified solver, and then transforms the solution into latent infections using the `sol2infs`
+function.
+
+NB: The `nothing` argument is a dummy latent process, e.g. a log-Rt time series, that is not
+used in the SIR model, but might be used in other models.
+
+```julia
+sir_infections_mdl = generate_latent_infs(sir_process, nothing)
+```
+
+We can generate some test data from the model by sampling from the model and then calling the model
+
+```julia
+# Sampled parameters
+θ = rand(sir_infections_mdl)
+expected_infections = (sir_infections_mdl | θ)()
+test_data = expected_infections .|> λ -> rand(Poisson(λ))
+```
+
+Now, we can refit the model using a `PoissonError` observation model
+
+```julia
+pois_obs = PoissonError()
+
+@model function fit_sir_model(data)
+    @submodel I_t = generate_latent_infs(sir_process, nothing)
+    @submodel _ = generate_observations(pois_obs, data, I_t)
+
+    return nothing
+end
+
+inference_mdl = fit_sir_model(test_data)
+chn = sample(inference_mdl, NUTS(), 2_000)
+summarize(chn)
+```
+
+We can compare the summarized chain to the sampled parameters to see that the model is fitting
+the data well.
+
+# Custom ODE models
+
+To define a custom ODE model, you need to define:
+
+- Some `CustomParams <: AbstractTuringParamModel` struct (the name doesn't need to be `CustomParams`!)
```

Comment:
Is this out of date? If so highlights danger in using julia vs example

Reply:
Redone this, and its covered by a `jldoctest` to flag if we move away from this custom pattern